Upgrade insteonplm to 0.8.2 (required refactoring)

teharris1 opened the pull request:
## Description:
*Breaking change* If you have created platform overrides in your configuration.yaml file to change a your INSTEON device to map to a different Home Assistant platform, that mapping will no longer be in effect. Please see the new device override capabilities in the [insteon_plm documentation][docs].

Significant upgrade to allow more complex devices to be created and managed. Significant improvement to message handling, especially the fixing of a race issue that causes the system to hang
Wide support for switch and dimmer devices (cat 0x01 and cat 0x02)
Additional complex devices added:
- FanLinc
- I/O Linc (was there but now a bit cleaner)
- On/Off outlet both top and bottom outlets controllable
- Smokebridge (unconfirmed)

**Related issue (if applicable):** 

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#4695

## Example entry for `configuration.yaml` (if applicable):
```yaml
# insteon_plm supported configuration variables
insteon_plm:
  port: SERIAL_PORT
  device_override:
     - address: INSTEON_ADDRESS
       cat: DEVICE_CATEGORY
       subcat: DEVICE_SUBCATEGORY
       firmware: DEVICE_FIRMWARE
       product_key: DEVICE_PRODUCT_KEY
```

## Checklist:
  - [x] The code change is tested and works locally.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/teharris1/home-assistant/blob/dev/homeassistant/components/insteon_plm.py#L18
[ex-import]: https://github.com/teharris1/home-assistant/blob/dev/homeassistant/components/insteon_plm.py#L38
[docs]: https://home-assistant.io/components/insteon_plm/

Commit:
Merge from current dev

Commit:
Update for Sensor approach

Commit:
Update reference to state classes

Commit:
Reference stateKey correctly

Commit:
Reference stateKey

Commit:
Change deviceInfo to a dict

Commit:
Pass state to properties method

Commit:
Add state info to device_state_attributes

Commit:
Update entity name to include state name

Commit:
Update for on() off() vs light_on/off

Commit:
Flag newnames option

Commit:
Update configuration schema

Commit:
Update configuration schema

Commit:
Spell False correctly

Commit:
Rename state to statekey

Commit:
Rename statekey to stateKey

Commit:
Call new device with stateKey and newname

Commit:
Simplify use of newnames

Commit:
Add workdir to save devices

Commit:
Fix newnames config setup

Commit:
Propogate OnOffSensor to VariableSensor change

Commit:
Upgrade insteonplm version to 0.8.0

Commit:
Pass address rather than device object to platform

Commit:
Set inteon_plm data variable to PLM

Commit:
Consistant use of conn and plm

Commit:
Consistant use of conn and plm

Commit:
Proper reference to device and address

Commit:
Fixed platform setup issues

Commit:
Correct issue with missing _supported_features attr

Commit:
Properly reference self._state vs self.state

Commit:
Bump insteonplm version to 0.8.1

Commit:
Merge from current HA dev branch

Commit:
Remove subplatform and map based on state name

Commit:
Correct refrence to State

Commit:
Correct reference to device.states

Commit:
Bump insteonplm to 0.8.2

Commit:
Merge with latest dev branch

Commit:
Fix format errors

Commit:
Fix format issues

Commit:
Fix trailing whitespace

Commit:
Correct format issues

Commit:
Fix format issues

Commit:
Fix format issues

Commit:
Fixed format issues

Commit:
Fixed format issues

Commit:
Move imports inside classes

Commit:
Simplify import of modules

Commit:
Correct reference to OnOffSwitch_OutletTop and bottom

Commit:
Remove unnessary references

Commit:
Fix format issues

Commit:
Code review adjustments

Commit:
Fix format issue

Commit:
Use new nameing format for groups that are not group 0x01

Commit:
Remove newname feature

Commit:
Fix binary_sensor type to default to None

Commit:
Fix device_class property to return the sensor type correctly.

Commit:
rename _device and _state to avoid conflicts with Entity

Commit:
Format long lines

Commit:
Pylint clean up

Commit:
Insteon_PLM

Commit:
lint cleanup

Commit:
Check device_override has address

Commit:
3.4 lint clean up

Commit:
3.4 lint clean up

teharris1 commented:
This was originally posted in [PR 12483][pr-old]. I closed that PR because it added too much functionality while refactoring. This PR is focused on the refactory as much as possible to simplify the assessment of the PR. 

[pr-old]: https://github.com/home-assistant/home-assistant/pull/12483

MartinHjelmare commented on homeassistant/components/fan/insteon_plm.py:
```diff
@@ -0,0 +1,137 @@
+Support for INSTEON fans via PowerLinc Modem.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/insteon_plm/
```
The url is wrong. Please look at another platform for the standard format.

MartinHjelmare commented on homeassistant/components/fan/insteon_plm.py:
```diff
@@ -0,0 +1,137 @@
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up the INSTEON PLM device class for the hass platform."""
+    state_list = []
```
It's not a list of states. Please change the variable name to something like `device_list` or `entities`.

MartinHjelmare commented on homeassistant/components/fan/insteon_plm.py:
```diff
@@ -0,0 +1,137 @@
+class InsteonPLMFan(FanEntity):
+    """An INSTEON fan component."""
+
+    def __init__(self, hass, device, state_key,
```
Don't pass in `hass`.

MartinHjelmare commented on homeassistant/components/fan/insteon_plm.py:
```diff
@@ -0,0 +1,137 @@
+        self._insteon_device = device
+        self._supported_features = supported_features
+
+        self._insteon_device_state.register_updates(self.async_fan_update)
```
Move to `async_added_to_hass`.

MartinHjelmare commented on homeassistant/components/fan/insteon_plm.py:
```diff
@@ -0,0 +1,137 @@
+    @property
+    def speed_list(self) -> list:
+        """Get the list of available speeds."""
+        return [STATE_OFF, SPEED_LOW, SPEED_MEDIUM, SPEED_HIGH]
```
Please define this list as a constant at the module level and return that here.

MartinHjelmare commented on homeassistant/components/insteon_plm.py:
```diff
@@ -6,16 +6,16 @@
 from homeassistant.core import callback
-from homeassistant.const import (
-    CONF_PORT, EVENT_HOMEASSISTANT_STOP)
+from homeassistant.const import (CONF_PORT, EVENT_HOMEASSISTANT_STOP)
```
Don't use parenthesis if you don't break the line.

MartinHjelmare commented on homeassistant/components/insteon_plm.py:
```diff
@@ -26,95 +26,142 @@
+                hass.async_add_job(
+                    discovery.async_load_platform(
+                        hass, platform, DOMAIN,
+                        discovered=[{'address': device.address.hex,
```
Discovery info is usually a dictionary. I'm not sure if it's required to be that. Why should it be a list?

teharris1 replied:
This is a very interesting question. I left it this way because that is the way it was in the prior version of this code but looking at it, there does not seem to be a good reason for it. However, it is logical given that the callback method of async_setup_platform takes a list so it is logical to assume that discovered is a list. But you are right, it does not have to be since I am only calling one entity at a time to add to the platform. I will test it out and see if anything breaks :)

MartinHjelmare commented on homeassistant/components/insteon_plm.py:
```diff
@@ -26,95 +26,142 @@
-        else:
-            plm.protocol.devices.add_override(
-                device['address'], 'capabilities', [device['platform']])
+        address = device_override.get('address', None)
```
`None` is the default value returned if the key is missing.

teharris1 replied:
Yes, this is by design. Providing fault tolerance for a user that may not correctly define the configuration correctly. If no address is defined in the configuration then the device_override should not be passed to the underlying library. But as you point this out, if an address is not defined an error should be logged. So I will notify the user with an error message that the configuration is incorrect.

MartinHjelmare replied:
Please remove it.

MartinHjelmare replied:
This still needs addressing too.

MartinHjelmare commented on homeassistant/components/light/insteon_plm.py:
```diff
@@ -22,36 +22,31 @@
 @asyncio.coroutine
 def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
     """Set up the Insteon PLM device."""
+    state_list = []
```
Please change the variable name.

MartinHjelmare commented on homeassistant/components/light/insteon_plm.py:
```diff
@@ -22,36 +22,31 @@
     """A Class for an Insteon device."""
 
-    def __init__(self, hass, plm, address, name, dimmable):
+    def __init__(self, hass, device, state_key):
```
Don't pass in `hass`.

MartinHjelmare commented on homeassistant/components/light/insteon_plm.py:
```diff
@@ -22,36 +22,31 @@
-    def __init__(self, hass, plm, address, name, dimmable):
+    def __init__(self, hass, device, state_key):
         """Initialize the light."""
         self._hass = hass
```
Remove.

MartinHjelmare commented on homeassistant/components/light/insteon_plm.py:
```diff
@@ -22,36 +22,31 @@
 
-        self._plm.add_update_callback(
-            self.async_light_update, {'address': self._address})
+        self._insteon_device_state.register_updates(self.async_light_update)
```
Move to `async_added_to_hass`.

MartinHjelmare commented on homeassistant/components/insteon_plm.py:
```diff
@@ -26,95 +26,142 @@
+        'Model': entity.model,
+        'Category': '{:02x}'.format(entity.cat),
+        'Subcategory': '{:02x}'.format(entity.subcat),
+        'Product Key / Firmware': '{:02x}'.format(entity.product_key),
```
Firmware version and similar info is not allowed to be part of state attributes. The state machine should only hold attributes that describe the state of the entities.

teharris1 replied:
I made this change. Just so you know, this is meaningfully the same as it was in the original code just using the refactored approach to the insteonplm module. I see a lot of other platforms using the state machine the same way we were. I do think there is value to the user to be able to click on the entity and see the manufacturer, model number, etc. As more and more users start using HA we can expect they become less and less technical. Having that easily accessible I believe makes the platform more user friendly at a very low cost. (My two cents for what it is worth but will respect your direction.)

MartinHjelmare commented on homeassistant/components/light/insteon_plm.py:
```diff
@@ -61,59 +56,57 @@ def should_poll(self):
+    def async_light_update(self, entity_id, statename, val):
         """Receive notification from transport that new data exists."""
-        _LOGGER.info("Received update callback from PLM for %s", self._address)
         self._hass.async_add_job(self.async_update_ha_state())
```
`self.async_schedule_update_ha_state()`

MartinHjelmare commented on homeassistant/components/sensor/insteon_plm.py:
```diff
@@ -0,0 +1,85 @@
+Support for INSTEON dimmers via PowerLinc Modem.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/insteon_plm/
```
Wrong url.

MartinHjelmare commented on homeassistant/components/sensor/insteon_plm.py:
```diff
@@ -0,0 +1,85 @@
+https://home-assistant.io/components/insteon_plm/
+"""
+import logging
+import asyncio
```
Please sort 🔤 within groups standard library, 3rd party and homeassistant.

MartinHjelmare commented on homeassistant/components/sensor/insteon_plm.py:
```diff
@@ -0,0 +1,85 @@
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up the INSTEON PLM device class for the hass platform."""
+    state_list = []
```
Please change variable name.

MartinHjelmare commented on homeassistant/components/sensor/insteon_plm.py:
```diff
@@ -0,0 +1,85 @@
+class InsteonPLMSensorDevice(Entity):
+    """A Class for an Insteon device."""
+
+    def __init__(self, hass, device, state_key):
```
Don't pass in `hass`.

MartinHjelmare commented on homeassistant/components/sensor/insteon_plm.py:
```diff
@@ -0,0 +1,85 @@
+        self._insteon_device_state = device.states[state_key]
+        self._insteon_device = device
+
+        self._insteon_device_state.register_updates(self.async_sensor_update)
```
Move to `async_added_to_hass`.

MartinHjelmare commented on homeassistant/components/sensor/insteon_plm.py:
```diff
@@ -0,0 +1,85 @@
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        sensorstate = self._insteon_device_state.value
```
You can return the attribute directly.

MartinHjelmare commented on homeassistant/components/sensor/insteon_plm.py:
```diff
@@ -0,0 +1,85 @@
+    @callback
+    def async_sensor_update(self, deviceid, statename, val):
+        """Receive notification from transport that new data exists."""
+        self._hass.async_add_job(self.async_update_ha_state())
```
`self.async_schedule_update_ha_state()`

MartinHjelmare commented on homeassistant/components/switch/insteon_plm.py:
```diff
@@ -19,34 +19,38 @@
 @asyncio.coroutine
 def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
     """Set up the INSTEON PLM device class for the hass platform."""
+    state_list = []
```
Please revert variable name change.

MartinHjelmare replied:
Use `entities` or `entity_list`.

MartinHjelmare commented on homeassistant/components/switch/insteon_plm.py:
```diff
@@ -19,34 +19,38 @@
     """A Class for an Insteon device."""
 
-    def __init__(self, hass, plm, address, name):
+    def __init__(self, hass, device, state_key):
```
Don't pass in `hass`.

MartinHjelmare commented on homeassistant/components/switch/insteon_plm.py:
```diff
@@ -19,34 +19,38 @@
-    def __init__(self, hass, plm, address, name):
+    def __init__(self, hass, device, state_key):
         """Initialize the switch."""
         self._hass = hass
```
Remove.

MartinHjelmare commented on homeassistant/components/switch/insteon_plm.py:
```diff
@@ -19,34 +19,38 @@
 
-        self._plm.add_update_callback(
-            self.async_switch_update, {'address': self._address})
+        self._insteon_device_state.register_updates(self.async_switch_update)
```
Move to `async_added_to_hass`.

MartinHjelmare commented on homeassistant/components/switch/insteon_plm.py:
```diff
@@ -56,42 +60,104 @@ def should_poll(self):
+    @callback
+    def async_switch_update(self, deviceid, statename, val):
+        """Receive notification from transport that new data exists."""
+        self._hass.async_add_job(self.async_update_ha_state())
```
`self.async_schedule_update_ha_state()`

MartinHjelmare commented on homeassistant/components/switch/insteon_plm.py:
```diff
@@ -56,42 +60,104 @@ def should_poll(self):
+class InsteonPLMOpenClosedDevice(SwitchDevice):
+    """A Class for an Insteon device."""
+
+    def __init__(self, hass, device, state_key):
```
Don't pass in `hass`.

MartinHjelmare commented on homeassistant/components/switch/insteon_plm.py:
```diff
@@ -56,42 +60,104 @@ def should_poll(self):
+        self._insteon_device_state = device.states[state_key]
+        self._insteon_device = device
+
+        self._insteon_device_state.register_updates(self.async_relay_update)
```
Move to `async_added_to_hass`.

MartinHjelmare commented on homeassistant/components/switch/insteon_plm.py:
```diff
@@ -56,42 +60,104 @@ def should_poll(self):
+    def async_relay_update(self, deviceid, statename, val):
         """Receive notification from transport that new data exists."""
-        _LOGGER.info('Received update callback from PLM for %s', self._address)
         self._hass.async_add_job(self.async_update_ha_state())
```
`self.async_schedule_update_ha_state()`

MartinHjelmare commented on homeassistant/components/fan/insteon_plm.py:
```diff
@@ -0,0 +1,137 @@
+    @callback
+    def async_fan_update(self, deviceid, statename, val):
+        """Receive notification from transport that new data exists."""
+        self.hass.async_add_job(self.async_update_ha_state())
```
`self.async_schedule_update_ha_state()`

MartinHjelmare commented on homeassistant/components/binary_sensor/insteon_plm.py:
```diff
@@ -15,38 +15,43 @@
+        self._insteon_device_state = device.states[state_key]
+        self._insteon_device = device
+        self._sensor_type = SENSOR_TYPES.get(self._insteon_device_state.name,
+                                             None)
```
`None` is the default value returned if the key is missing.

teharris1 replied:
Yes, this is by design. If you look at the `BinarySensorDevice` base class, the default value of the `device_class` property is `None`. So if the underlying library produces a binary sensor that is not in one of the `SENSOR_TYPES` it should default to `None` and produce the default behavior of `BinarySensorDevice`.

MartinHjelmare replied:
Yeah, I just would like you to remove the explicit default of `None` here, cause it's already the default default for `dict.get`.

teharris1 commented on homeassistant/components/binary_sensor/insteon_plm.py:
```diff
@@ -15,38 +15,43 @@
 @asyncio.coroutine
 def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
     """Set up the INSTEON PLM device class for the hass platform."""
+    state_list = []
```
I am good to call it an entity. the reason it is changed from a device is because in INSTEON terms it is not a device, it is a state within the device. It is a terminology mapping issue but I am fine to map to the HA terminology of entity. FYI, this was one of the major changes to the underlying insteonplm library. Many complex devices (such as FanLinc) have two entities inside them (i.e. a light at a fan in the FanLinc case) From an INSTEON perspective they are one device with two controllable entities (called states in that library). So calling it a state_list is mapping to the insteonplm naming convention but I am fine to map to the HA naming convention of entity.

MartinHjelmare replied:
Good, please use entity convention inside home assistant.

Commit:
Changes per code review

Commit:
Change discovered from a list of dict to a dict

Commit:
Correct common_attributes usage

Commit:
Change discovered from a list of dict to a dict

Commit:
Add debugging message to confirm platform setup

Commit:
Debug messages

Commit:
Debug messages

Commit:
Debug async_added_to_hass

Commit:
Debug async_added_to_hass async_add_job

Commit:
Debug async_added_to_hass

Commit:
Debug devices not loading in hass

Commit:
Debug new entities not being added to hass

Commit:
Debug adding devices to hass

Commit:
Revert "3.4 lint clean up"

This reverts commit 0d8fb992b12e9eea86716753fd2f302b1addb458.

Commit:
Revert "Debug adding devices to hass"

This reverts commit ec306773d47401b100bcdaaf0af47c93699d78b4.

Commit:
Revert "Debug new entities not being added to hass"

This reverts commit 55fb724d06e7d1e249de46acb2de7eac2eb7d14d.

Commit:
Revert "Debug devices not loading in hass"

This reverts commit 07814b4f14cab85e67197812b055a2d71a954b1f.

Commit:
Revert "Debug async_added_to_hass"

This reverts commit 4963a255d86c1bf63ec6064b0d911893d310f13d.

Commit:
Revert "Debug async_added_to_hass async_add_job"

This reverts commit 22cadff91f524edf91605c4c1f9df0a3d125d1d9.

Commit:
Revert "Debug async_added_to_hass"

This reverts commit 12c5651fe497b439ba962473973232ae9745314d.

teharris1 commented:
@MartinHjelmare I believe I have made the changes you recommended. Happy to revisit any items you think were not complete. Thank you for your time and showing me some better ways to work with HA.

MartinHjelmare commented on homeassistant/components/binary_sensor/insteon_plm.py:
```diff
@@ -15,38 +15,39 @@
+        self._insteon_device_state = device.states[state_key]
+        self._insteon_device = device
+        self._sensor_type = SENSOR_TYPES.get(self._insteon_device_state.name,
+                                             None)
```
Please remove default `None`. It's already the default default value.

MartinHjelmare commented on homeassistant/components/fan/insteon_plm.py:
```diff
@@ -0,0 +1,147 @@
+        """Get the list of available speeds."""
+        return FAN_SPEEDS
+
+    def async_turn_on(self, speed: str = None, **kwargs) -> None:
```
The methods should be annotated with `@asyncio.coroutine`.

MartinHjelmare commented on homeassistant/components/fan/insteon_plm.py:
```diff
@@ -0,0 +1,147 @@
+
+    def async_turn_off(self, **kwargs) -> None:
+        """Turn off the entity."""
+        self.async_set_speed(SPEED_OFF)
```
This will be a coroutine so `yield from` it.

MartinHjelmare commented on homeassistant/components/fan/insteon_plm.py:
```diff
@@ -0,0 +1,147 @@
+        """Turn on the entity."""
+        if speed is None:
+            speed = SPEED_MEDIUM
+        self.async_set_speed(speed)
```
`yield from` this coroutine.

MartinHjelmare replied:
This needs to be addressed.

MartinHjelmare commented on homeassistant/components/binary_sensor/insteon_plm.py:
```diff
@@ -56,32 +57,51 @@ def should_poll(self):
+        _LOGGER.debug('Device %s entity %s added. Now registering callback.',
+                      self.address, self._insteon_device_state.name)
+        self.hass.async_add_job(
+            self._insteon_device_state.register_updates,
```
If this just appends the callback to a list, that is async safe, and then you do not need to use `async_add_job`.

MartinHjelmare commented on homeassistant/components/binary_sensor/insteon_plm.py:
```diff
@@ -56,32 +57,51 @@ def should_poll(self):
-    def name(self):
-        """Return the name of the node."""
-        return self._name
+    def group(self):
```
Where is this used?

teharris1 replied:
It is being used in the state machine attributes. I removed the manufacturer, model number, etc per you request above and only left two items, address and group. These are the minimum attributes to describe the unique entity of the physical device.

teharris1 commented:
@MartinHjelmare Thank you for pointing out the potential async calls. I did have a question about the Travis CI failures. I cannot see any connection between this PR and the code that is failing in the CI build. Any direction you can provide would be appreciated. Thanks.

teharris1 commented:
@MartinHjelmare disregard my Travis CI question. I am now seeing that the async methods were the issue since they all now pass. Feedback to the test writer that the test provides no feedback about the nature of the issue (i.e. the issue was in fan/insteon_plm.py but there is no way to know that from any of the message outputs). All is good now so thanks for seeing my error.

MartinHjelmare commented on homeassistant/components/fan/insteon_plm.py:
```diff
@@ -0,0 +1,147 @@
+        """Turn off the entity."""
+        yield from self.async_set_speed(SPEED_OFF)
+
+    def async_set_speed(self, speed: str) -> None:
```
This should also be annotated with `@asyncio.coroutine`.

MartinHjelmare reviewed (commented):
I think there are two remaining comments to address.

Another think that could save a lot of code is to move all common entity methods and properties to a super class that inherits from `Entity`. Then use multiple inheritance and inherit from this super class in the platform entity classes, besides the platform entity base class.

I think you can move `__init__`, `group`, `address`, `should_poll`, `name`, `device_state_attributes`, `async_added_to_hass`, and a common update callback to the super class.

teharris1 commented:
@MartinHjelmare I created the base class and corrected the missing coroutine decorator. I have updated the documentation as well. It appears that requires a review too. I want to make sure they both go together. Can you help me understand the process and is there anything I need to do to make sure both changes go into effect at the same time? I did identify this PR in the documentation PR and vis versa.

MartinHjelmare commented on homeassistant/components/binary_sensor/insteon_plm.py:
```diff
@@ -2,86 +2,66 @@
-        """Return the address of the node."""
-        return self._address
+    @callback
+    def async_binarysensor_update(self, deviceid, statename, val):
```
You can move this method to the common base class too. Just rename it to something general like `update_callback`.

MartinHjelmare commented on homeassistant/components/binary_sensor/insteon_plm.py:
```diff
@@ -2,86 +2,66 @@
-        _LOGGER.info("Received update callback from PLM for %s", self._address)
-        self._hass.async_add_job(self.async_update_ha_state())
+    @asyncio.coroutine
+    def async_added_to_hass(self):
```
As above, this method can be moved to the common base class.

MartinHjelmare reviewed (changes requested):
Nice!

Please also replace the update callback and `async_added_to_hass` in all platforms with just two methods in the common base class `InsteonPLMEntity`. The update callbacks don't need to have different names, like they have now.

Regarding the docs PR, it's enough to link from both PRs as you've done, and then just make sure you target the `next` branch in the docs PR. That will make sure the PR gets into the next release after both PRs have been merged.

MartinHjelmare reviewed (approved):
Looks great!

MartinHjelmare commented:
Fix the lint issues, and we can merge this. :+1:

Commit:
Pylint clean up

MartinHjelmare commented:
I think you might need to squash some commits to get the CLA bot to comply. It doesn't like to operate on more than 100 commits per PR.

Commit:
pylint cleanup

Commit:
Clean up naming

MartinHjelmare commented on homeassistant/components/binary_sensor/insteon_plm.py:
```diff
@@ -2,86 +2,56 @@
-        return self._address
+    def __init__(self, device, state_key):
+        """Initialize the INSTEON PLM binary sensor."""
+        InsteonPLMEntity.__init__(self, device, state_key)
```
```py
super().__init__(device, state_key)
```

MartinHjelmare commented on homeassistant/components/binary_sensor/insteon_plm.py:
```diff
@@ -2,86 +2,56 @@
+    state_key = discovery_info['state_key']
 
-    async_add_devices(device_list)
+    _LOGGER.debug('Adding device %s entity %s to Binary Sensor platform.',
```
Don't end the log message with a period.

MartinHjelmare commented on homeassistant/components/fan/insteon_plm.py:
```diff
@@ -0,0 +1,96 @@
+    device = plm.devices[address]
+    state_key = discovery_info['state_key']
+
+    _LOGGER.debug('Adding device %s entity %s to Fan platform.',
```
Remove period at the end.

MartinHjelmare commented on homeassistant/components/fan/insteon_plm.py:
```diff
@@ -0,0 +1,96 @@
+    @asyncio.coroutine
+    def async_turn_off(self, **kwargs) -> None:
+        """Turn off the entity."""
+        self.async_set_speed(SPEED_OFF)
```
`yield from self.async_set_speed(...)`

teharris1 replied:
Somewhat related question. `async_setup_platform`  passes in an add device handler (in this code it is called `async_add_devices`. Should that be `yield from`?

MartinHjelmare replied:
No, `async_add_devices` is not a coroutine.

MartinHjelmare commented on homeassistant/components/insteon_plm.py:
```diff
@@ -26,95 +26,183 @@
-        vol.Optional(CONF_OVERRIDE, default=[]): cv.ensure_list_csv,
-    })
+        vol.Optional(CONF_OVERRIDE): vol.All(
+            cv.ensure_list_csv, vol.Length(min=1))
```
Add a config schema for the override dict. It should validate address, cat, subcat, firmware and product key. Remove the `vol.Length` validator.

teharris1 replied:
So if I am going to make this modification I am thinking about adding a check if the old configuration option is still in use (this is the breaking change noted above). If the old configuration is in place, I am considering putting a `persistent_notification` to let the user know that the configuration options have changed. Is this considered the correct use of notifications?

MartinHjelmare replied:
You can mark it as deprecated via that validator. It's in the config_validation helper. It will log it. I wouldn't do a persistent notification.

teharris1 replied:
Holy crap, this was tough because the documentation for voluptuous is very sparse. But it was definitely the right thing to do. I tested it pretty well so I feel good about it. Thanks for the recommendation.

MartinHjelmare commented on homeassistant/components/light/insteon_plm.py:
```diff
@@ -24,96 +23,47 @@ def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    state_key = discovery_info['state_key']
 
-        _LOGGER.info("Registered %s with light platform", name)
+    _LOGGER.debug('Adding device %s entity %s to Light platform.',
```
Remove period at the end.

MartinHjelmare commented on homeassistant/components/sensor/insteon_plm.py:
```diff
@@ -0,0 +1,36 @@
+    device = plm.devices[address]
+    state_key = discovery_info['state_key']
+
+    _LOGGER.debug('Adding device %s entity %s to Sensor platform.',
```
Same as above.

teharris1 replied:
I applied this to all the platforms :)

MartinHjelmare commented on homeassistant/components/switch/insteon_plm.py:
```diff
@@ -21,77 +20,54 @@ def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
 
-class InsteonPLMSwitchDevice(SwitchDevice):
-    """A Class for an Insteon device."""
+    _LOGGER.debug('Adding device %s entity %s to Switch platform.',
```
Same as above.

MartinHjelmare commented on homeassistant/components/insteon_plm.py:
```diff
@@ -26,95 +26,183 @@
-            plm.protocol.devices.add_override(
-                device['address'], 'capabilities', [device['platform']])
+        address = device_override.get('address')
+        if address is not None:
```
After adding the override config validation schema, this check can be removed.

MartinHjelmare commented on homeassistant/components/insteon_plm.py:
```diff
@@ -26,95 +26,183 @@
+        address = device_override.get('address')
+        if address is not None:
+            if device_override.get('cat', False):
+                plm.devices.add_override(address,
```
Maybe do this with a `for` loop, iterating over the four optional keys of the override. If there is a value for the key in the override, add the override.

MartinHjelmare commented on homeassistant/components/insteon_plm.py:
```diff
@@ -26,95 +26,183 @@
+                                         device_override['subcat'])
+            if device_override.get('firmware', False):
+                plm.devices.add_override(address,
+                                         'product_key',
```
Is it correct to use `'product_key' ` here? That doesn't match the `'firmware' ` key.

teharris1 replied:
It is. They are used interchangeably in the underlying module unfortunately. It is one of the things that has to be changed there.

MartinHjelmare reviewed (changes requested):
I had another look and noticed some things that should be changed. Sorry that I missed them before.

Commit:
Enhance config schema. Fix logging issue

MartinHjelmare commented on homeassistant/components/insteon_plm.py:
```diff
@@ -22,12 +22,28 @@
+CONF_SUBCAT = 'subcat'
+CONF_FIRMWARE = 'firmware'
+CONF_PRODUCT_KEY = 'product_key'
+CONF_PLATFORM = "platform"
```
This can be imported from const.py.

MartinHjelmare commented on homeassistant/components/insteon_plm.py:
```diff
@@ -4,117 +4,211 @@
-                device['address'], 'capabilities', [device['platform']])
+        address = device_override.get('address')
+        for prop in device_override:
+            if prop in [CONF_ADDRESS, CONF_CAT, CONF_SUBCAT]:
```
Can you override the address too? This wasn't in the previous version, right?

MartinHjelmare reviewed (approved):
Looks good!

teharris1 commented:
@MartinHjelmare Thank you very much for all your help. Do I assume I need to squash commits again but what is the deal with the Travis CI build? The build fails in 3.6 `TestHomeKit.test_homekit_pyhap_interaction`. I cannot figure out why my build would fail on this test. Any ideas?

MartinHjelmare commented:
Yeah, I think you need to squash again to let the CLA bot finish. The failed test is probably a flaky test. Don't worry about it. Hopefully it'll pass on the next build.

Commit:
Reapply changes after squash

teharris1 commented:
@MartinHjelmare Sqash done so CLA working now. 3.6 is passing now. CI Test failing now in 3.5 test set 3 `testpilight`. I assume this has nothing to do with my code so hoping you can approve it to go forward.

teharris1 commented:
@MartinHjelmare Looks like all tests have passed! THANK YOU!. I really appreciate all the effort you put in to this. This is a much better code base to work from going forward because of your efforts. It is people like you that will make this HA platform successful. Thanks again.

MartinHjelmare commented:
Appreciate the feedback, thanks! Great work on the PR! :tada:
